[syntax-errors] Tuple unpacking in `return` and `yield` before Python 3.8

ntBre opened the pull request:
Summary
--

Checks for tuple unpacking in `return` and `yield` statements before Python 3.8, as described [here].

Test Plan
--
Inline tests.

[here]: https://github.com/python/cpython/issues/76298

MichaReiser commented on crates/ruff_python_parser/src/parser/statement.rs:
```diff
@@ -388,10 +396,29 @@ impl<'src> Parser<'src> {
+            // # parse_options: {"target-version": "3.7"}
+            // rest = (4, 5, 6)
+            // def f(): return (1, 2, 3, *rest)
+            // def g(): yield (1, 2, 3, *rest)
```
Are the `yield` examples relevant to this code? It seems to me that this parse rule is only used for `return x` or `return yield`.

ntBre replied:
No, you're right. I was getting alarmed at the size of the diffs in these PRs and saw that grouping tests together made them smaller, but conceptually they are separate tests. I'll move those to the `yield` section.

dhruvmanila commented on crates/ruff_python_parser/src/parser/statement.rs:
```diff
@@ -388,10 +396,29 @@ impl<'src> Parser<'src> {
-            )
+            let parsed_expr = self.parse_expression_list(ExpressionContext::starred_bitwise_or());
+
+            // test_ok iter_unpack_return_py37
```
Can we add test cases using single unpack element like `return *rest` or `yield *rest`?

dhruvmanila replied:
What about `yield 1, (*rest)`, `yield 1, (yield 2, *rest), 3` (yield can be used as an expression)? I'm just throwing out random examples :)

ntBre replied:
Thanks for these! The single cases are actually still invalid on my system Python (3.13):

```pycon
>>> def foo(): return *rest
  File "<python-input-0>", line 1
    def foo(): return *rest
                      ^^^^^
SyntaxError: can't use starred expression here
>>> def foo(): yield *rest
  File "<python-input-1>", line 1
    def foo(): yield *rest
                     ^^^^^
SyntaxError: can't use starred expression here
```

But these aren't flagged as `ParseError`s by ruff.

The second case is very interesting. It's actually accepted on both 3.7 *and* 3.8, but not on my 3.13 system Python (`SyntaxError: cannot use starred expression here`). That might be another change not currently tracked in #6591.

I added your third case as an example of a yield expression and moved the check into `parse_yield_expression` as @MichaReiser suggested.

dhruvmanila replied:
> But these aren't flagged as `ParseError`s by ruff.

That's because the grammar says it's allowed. I think it's caught by the CPython compiler and not the parser. You can see that `python -m ast test.py` wouldn't raise a syntax error here.

```
return_stmt:
    | 'return' [star_expressions] 

star_expressions:
    | star_expression (',' star_expression )+ [','] 
    | star_expression ',' 
    | star_expression

star_expression:
    | '*' bitwise_or 
    | expression
```

dhruvmanila replied:
Oops, missed the issue https://github.com/astral-sh/ruff/issues/16520

ntBre commented on crates/ruff_python_parser/src/error.rs:
```diff
@@ -457,6 +464,12 @@ impl Display for UnsupportedSyntaxError {
+                "iterable unpacking in return statements"
+            }
+            UnsupportedSyntaxErrorKind::StarTuple(StarTupleKind::Yield) => {
+                "iterable unpacking in yield statements"
```
I'll have to test this out. The CPython PR said statements, but maybe I took it too literally.

ntBre commented:
I added more test cases (thanks for the suggestions!). Two of them became TODOs because I think they aren't exactly related to this PR:
* `return|yield *rest` is invalid on every Python version I tried and should be flagged as a *parse* error, in my opinion
* `return|yield 1, (*rest)` is valid on 3.7 and 3.8 but not 3.13, so I think this is a separate change from the 3.8 unpacking (and a second `Change::Removed` case to use the enum from the other PR)

I also checked on `yield from` and it looks like our [`parse_yield_from`](https://github.com/astral-sh/ruff/blob/d0623888b3231fdda09bfa7079d67f8162faa10c/crates/ruff_python_parser/src/parser/expression.rs#L2110) method already reports an error for any unparenthesized tuple expression, so I think that should be covered.

Finally, I changed the message for the `yield statement` to `yield expression`. I think in an ideal case we'd have a third variant of `StarTupleKind` and differentiate `YieldExpr` from `YieldStmt` but I didn't see an easy way to do that with the new detection in `parse_yield_expression`. Since I had to pick one, I thought "expression" was slightly preferable.

MichaReiser commented on crates/ruff_python_parser/src/parser/expression.rs:
```diff
@@ -2089,10 +2090,39 @@ impl<'src> Parser<'src> {
+            // def f(): yield *rest
+
+            // test_err iter_unpack_yield_todo
+            // # TODO(brent) this is accepted by 3.7 and 3.8 but not 3.13 (at least)
+            // # looks like a separate change from the tuple unpacking checked here
+            // def i(): yield 1, (*rest)
+            self.check_tuple_unpacking(&parsed_expr, StarTupleKind::Yield);
```
If you haven't done so. Can you add it to the syntax check issue?

ntBre replied:
Added to #6591! It looks like this stopped parsing in 3.9

MichaReiser commented on crates/ruff_python_parser/src/parser/statement.rs:
```diff
@@ -389,10 +390,37 @@ impl<'src> Parser<'src> {
+            // rest = (4, 5, 6)
+            // def f(): return 1, 2, 3, *rest
+
+            // TODO(brent) this should be invalid (raise a `ParseError`, not just an
+            // `UnsupportedSyntaxError`) as it's unsupported by every Python version I've tried
+            // (3.7, 3.8, and 3.13)
+            // test_err iter_unpack_return
+            // # invalid on every Python version
+            // def f(): return *rest
```
Same: Let's remove this as it is a semantic error and not a parse error

dhruvmanila commented on crates/ruff_python_parser/resources/inline/err/iter_unpack_return_todo.py:
```diff
@@ -0,0 +1,3 @@
+# TODO(brent) this is accepted by 3.7 and 3.8 but not 3.13 (at least).
+# looks like a separate change from the tuple unpacking checked here
+def i(): return 1, (*rest)
```
Oh, interesting!
